Misc small improvements

Martchus opened the pull request:
See the particular commit messages

Fixes: [poo#70252](https://progress.opensuse.org/issues/70252)

Amrysliu reviewed (commented):
Thanks for dealing with the dependency issue. I introduced this problem, thanks for fixing it.
The code looks much better than before.

kalikiana commented on lib/OpenQA/WebAPI/Controller/API/V1/Search.pm:
```diff
@@ -107,11 +107,12 @@ sub query {
         foreach my $match (@lines) {
             my ($filename, $occurrence) = split(':', $match);
-            push(@results, {occurrence => $distri->basename . '/' . $filename, contents => $occurrence});
+            push(@results, {occurrence => "$basename/$filename", contents => $occurrence}) if $filename;
```
I don't understand the purpose of this. Each line is `filename:contents`. Unless perhaps you're seeing a side effect of `:` being parsed incorrectly if the line contains `:`, which I'm addressing in #3317 by adding a limit to the `split` call.

Martchus replied:
If there are no search results a single line break might be returned. In this case  `$match` is empty. In this case `$filename` is not defined leading to `Use of uninitialized $filename` as stated in the commit message.

(And yes, `split` apparently not just returns an empty string here, e.g.
```
  DB<5> ($filename, $foo) = split(':', '');
  DB<6> print('not defined') unless defined $filename;
not defined
```
)

perlpunk replied:
Then it might be even better to do:
```
next unless length $match;
```
before the split.

kalikiana replied:
Ah. And that's not even a proper failure, which explains why the tests didn't catch it. 

Although in that case I'd rather do a `strip` or something else to avoid the loop. Semantically no filename just makes no sense.

Martchus replied:
I guess I follow @perlpunk's suggestion. That seems to be the simplest way to achieve the correct behavior.

Martchus commented:
By the way, I'm aware that `are_parents_ok` impacts the performance a little bit. The impact is notable within the profiler so I'm trying to optimize it a little bit.

Martchus commented:
I've integrated the `are_parents_ok` function within the `dependencies` function so we only loop though the parents one time and I also tried to access the parent job as less as possible. The performance impact is a little bit less now and I suppose it is ok.

I also changed the code for the search route as suggested and fixed a few places jshint complained about.
